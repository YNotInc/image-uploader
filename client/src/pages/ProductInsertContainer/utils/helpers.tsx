let setImageParagraphTag = (para: Text, color: string, imageName: string, imageSize: string, imageWidth: number, imageHeight: number, previewCanvas: HTMLCanvasElement) => {
  // Set Paragraph
  para = document.createTextNode('File Name: ' + imageName + ', File Size: ' + imageSize + ', Width: ' + imageWidth + ', Height: ' + imageHeight);

  let listItem = document.createElement('li');
  listItem.setAttribute('id', 'li-id');


  // listItem.appendChild(lnBreak);
  listItem.appendChild(para);

  let orderedList = document.createElement('ol');
  previewCanvas.appendChild(orderedList);

  orderedList.appendChild(listItem);

  listItem.style.color = color;
  // remove canvas if image added
  removeItem('canvas', previewCanvas);
};

let removeItem = (element: string, previewCanvas: HTMLCanvasElement) => {
  /************************/
  if (previewCanvas) {
    let item = previewCanvas.getElementsByTagName(`${element}`);
    const isItemCreated = item.length ? true : false;
    console.log("item Length", item.length);
    console.log("is item created", isItemCreated);
    console.log("no item", item);
    if (isItemCreated) {
      previewCanvas.removeChild(item[0]);
    }
    /************************/
  }

};

let removeCanvas = (previewCanvas: HTMLCanvasElement) => {
  if (previewCanvas) {
    var canvasElement = previewCanvas.getElementsByTagName('canvas');
    if (canvasElement.length >= 1) {
      removeItem('canvas', previewCanvas);
      removeItem('ol', previewCanvas);
    }
  }

};


let setFileMessage = (tag: string, message: string) => {
  const errorTag = document.getElementById(tag)!;
  if (errorTag) {
    errorTag.innerHTML = message;
  }
};

let getFormattedFileSize = (files: FileList): string => {
  let curFiles = files;
  return returnFileSize(curFiles[0].size)!;
};

/*Returns formatted version of file size*/
let returnFileSize = (number: number) => {
  if (number < 1024) {
    return number + 'bytes'
  } else if (number >= 1024 && number < 1048576) {
    return (number / 1024).toFixed(1) + 'KB'
  } else if (number >= 1048576) {
    return (number / 1048576).toFixed(1) + 'MB'
  }
};

let appendImage = (img: HTMLImageElement, canvas: HTMLCanvasElement, previewCanvas: HTMLCanvasElement) => {
  img.onload = () => {
    // Set canvas dimension to match image
    canvas.width = img.width
    canvas.height = img.height
    // canvas.width = previewCanvas.width;
    // canvas.height = previewCanvas.height;
    var context: CanvasRenderingContext2D = canvas.getContext('2d')!;

    // Draw image to canvas centered
    // https://stackoverflow.com/questions/16317971/draw-images-on-in-the-middle-of-a-canvas
    context.drawImage(img, canvas.width / 2 - img.width / 2,
      canvas.height / 2 - img.height / 2);
    // previewCanvas.appendChild(canvas);
    canvas.setAttribute("class", "img-fluid");
    previewCanvas.prepend(canvas);
  };
};

let checkImageDimensions = (imageWidth: number, imageMin: number, imageHeight: number, imageMax: number) => {
  let areDimensionsValid = false;
  // let para = document.createTextNode('');

  // if dimensions valid
  if (imageWidth >= imageMin && imageWidth <= imageMax && imageHeight >= imageMin && imageHeight <= imageMax) {
    areDimensionsValid = true;
  }
  // dimensions not valid
  else {
    areDimensionsValid = false;
  }
  return areDimensionsValid;
};

let setFileSize = (areDimensionsValid: boolean, errorTag: string, acceptedMsg: string, unacceptedMsg: string, imageName: string, imageSize: string, files: FileList, previewCanvas: HTMLCanvasElement, imageWidth: number, imageHeight: number) => {
  let para = document.createTextNode('');
  if (areDimensionsValid) {
    // Set file size with Units:
    imageSize = getFormattedFileSize(files);
    setImageParagraphTag(para, 'black', imageName, imageSize, imageWidth, imageHeight, previewCanvas);
    setFileMessage(errorTag, acceptedMsg);
    let submitBtn: HTMLButtonElement = document.getElementById('submit-image') as HTMLButtonElement;
    submitBtn.disabled = false;
  }
  else {
    console.log("****DIMENSIONS NOT VALID****");
    const submitImgBtn: HTMLButtonElement = document.getElementById('submit-image') as HTMLButtonElement;
    submitImgBtn.disabled = true;

    // Set file size with Units:
    imageSize = getFormattedFileSize(files);
    setImageParagraphTag(para, 'red', imageName, imageSize, imageWidth, imageHeight, previewCanvas);
    setFileMessage(errorTag, unacceptedMsg);
  }
  return imageSize;
};

let isFileSelected = (input: HTMLInputElement) => {
  let curFiles = input ? input.files : null;
  return curFiles;
};

let imgOnError = (previewCanvas: HTMLCanvasElement, imageWidth: number, imageMax: number, imageHeight: number, errorTag: string, invalidMsg: string) => {
  if (imageWidth <= imageMax && imageHeight <= imageMax) {
    console.log('NOT A Valid File: ');
    setFileMessage(errorTag, invalidMsg);
    const submitImgBtn: HTMLButtonElement = document.getElementById('submit-image') as HTMLButtonElement;
    submitImgBtn.disabled = true;
    console.log("IMAGE ERROR ONLOAD");
  }
  // Remove Canvas and Paragraph for wrong dimensions or no file.
  removeCanvas(previewCanvas);
  console.log("in imgONERROR");
}; // oneerror


let displayImage = (img: HTMLImageElement, areDimensionsValid: boolean, previewCanvas: HTMLCanvasElement) => {
  var canvasElement = previewCanvas.getElementsByTagName('canvas');
  var canvas = document.createElement('canvas');
  console.log("DISPLAYIMAGE CANVASELEMENT:", canvasElement);
  if (!areDimensionsValid) {

  }
  // No Image added to canvas
  else if (canvasElement.length === 0) {
    appendImage(img, canvas, previewCanvas)
  }
  // Image added to canvas
  else {
    removeItem('canvas', previewCanvas);
    appendImage(img, canvas, previewCanvas);
  }
};

let loadImage = (img: HTMLImageElement, blob: Blob): Promise<ImagePropsType> => {
  console.log("IMG", img, "BLOB", blob);
  return new Promise(function (resolve, reject) {
    // define source
    img.src = URL.createObjectURL(blob);
    // resolve promise onLoad
    img.onload = () => {

      const imageProps: ImagePropsType = {
        imageName: blob.name,
        imageHeight: img.height,
        imageWidth: img.width,
        imageSize: blob.size,
        imageSrc: img.src
      };
      resolve(imageProps);
    };
    // reject promise onError
    img.onerror = () => {
      reject("rejected");
    };
  });
};

/* Base64 Decoder: Remove the metadata
https://www.base64decode.net/base64-image-decoder */
let convertImageFromUrlToBase64String = async (url: string): Promise<string> => {
  console.log("In convertImageStart");
  var img = new Image();
  img.crossOrigin = 'Anonymous'
  var dataUrl;
  return new Promise(function (resolve, reject) {
    // Setting the img.src will call img.onload when the src is loaded
    console.log("URL1: ", url);
    // Pull Width/Height from URL and store in img objec
    img.src = url;// url is the img src
    console.log("URL2: ", url);

    img.onload = () => {
      var canvas = document.createElement('canvas');
      console.log(`canvas width: ${canvas.width}`);
      canvas.width = img.width;
      canvas.height = img.height;
      console.log("IMGWIDTH:", img.width, "IMGHEIGHT:", img.height);

      // Get a canvas reference to draw to the canvas
      var context: CanvasRenderingContext2D = canvas.getContext('2d')!;
      console.log(`canvas width: ${canvas.width}`);
      // Draw image to the canvas centered
      // https://stackoverflow.com/questions/16317971/draw-images-on-in-the-middle-of-a-canvas
      context.drawImage(img, canvas.width / 2 - img.width / 2,
        canvas.height / 2 - img.height / 2);
      // Return a data URI containing a representation of the image in jpg format
      dataUrl = canvas.toDataURL('image/jpg');
      console.log("Images loaded helper.js resolved", typeof dataUrl);
      resolve(dataUrl);
    };

    // reject promis onError
    img.onerror = (err) => {
      reject("image rejected in BASE64 Conversion:"+ err);
    };
  }); // promise
};// convertImage

// Check whether the file type of the input file is valid
let isFileTypeValid = (file: File, fileTypes: string[]) => {
  if (file) {
    console.log("FILEType :", file.type);
    for (var i = 0; i < fileTypes.length; i++) {
      if (file.type === fileTypes[i]) {
        return true;
      }
    }
  }
  else
    return false;
};

export {
  setImageParagraphTag, removeItem, removeCanvas, setFileMessage, getFormattedFileSize, appendImage, checkImageDimensions, setFileSize, isFileSelected, imgOnError, displayImage,
  loadImage, convertImageFromUrlToBase64String, isFileTypeValid, returnFileSize
};