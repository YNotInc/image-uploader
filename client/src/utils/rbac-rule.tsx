const rules = {
    visitor: {
      static: [
        "products:hide",
        "products:view",
        "home-page:visit"
        ]
    },
    user : {
      static: [ 
        "products:view", 
        "products:details",
          "products:hide",
          "home-page:visit"
      ]
    },
    admin: {
      static: [
        "products:view",
        "products:hide",
        "products:update",
        "products:insert",
        "products:delete",
        "products:details",
        "users:get",
        "users:getSelf",
        "home-page:visit",
        "dashboard-page:visit"
      ]
    },
    writer: {
      static: [
        "posts:list",
        "posts:create",
        "users:getSelf",
        "home-page:visit",
        "dashboard-page:visit"
      ],  //static
      // dynamic not currently used
      dynamic: {
        "posts:edit": (userId:string , postOwnerId: string) => {
          if (!userId || !postOwnerId) return false;
          return userId === postOwnerId;
        }
      } // dyanmic
    }
  };

  export default rules;