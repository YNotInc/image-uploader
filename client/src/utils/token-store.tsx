// Import Server-Side Utilities:
import API from './API';

export default {
    refresh: async (url:string, accessToken: string, refreshToken: string, email: string, expired:boolean) => {
        console.log('In REFRESHTOKENS:', refreshToken);
        /*************************************************************
         *  package and send the body to the endpoint
         ************************************************************/
        let newCredentialResponse = await API.refreshTokens(url, accessToken, refreshToken, email, expired);
        console.log("tokenStore: newCredentialResponse = ", newCredentialResponse);

        /*************************************************************
         * Send the newCredentialResponse back to the calling program
         ************************************************************/
        return newCredentialResponse;
    }
};