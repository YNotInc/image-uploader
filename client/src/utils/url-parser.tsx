export let urlParser = () => {
        let url = window.location.pathname;
        let urlArray = url.split('/');
        console.log("@@@URLARRAY:", urlArray);
        return urlArray[urlArray.length-1];
        // const id = urlArray[urlArray.length - 1];
        // console.log("ID:", id);
        //replace space with app
        // urlArray.splice(0, 1, '/api');
        // remove product from url
        // URL Syntax: /api/products/:id
        // urlArray.splice(2, 1);
        // console.log("newURLElements:", urlArray);
        // const baseUrl = urlArray.join('/');
        // console.log("baseUrl:", baseUrl);
};