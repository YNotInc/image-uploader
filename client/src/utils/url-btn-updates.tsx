import { urlParser } from "./url-parser";

export let urlBtnUpdates = () => {
        const pageName = urlParser();
        console.log("PageName:", pageName);
        // Add a de-activate or activate class to the button with the corresponding name
        let btnElement;
        let className;
        let element;
        if (pageName === '' || pageName === 'home') {
                btnElement = 'home';
                className = 'activate';

                //set other btn to de-activate
                element = document.querySelector(`[name=products]`);

                element?.setAttribute('class', ' de-activate nav-link Ripple-parent');
        }
        else if (pageName === 'products') {
                btnElement = 'products';
                className = 'activate';

                //set other btn to de-activate
                element = document.querySelector(`[name=home]`);

                element?.setAttribute('class', ' de-activate nav-link Ripple-parent');
        }
        else {
                // reset both buttons

                element = document.querySelector(`[name=home]`);

                element?.setAttribute('class', ' de-activate nav-link Ripple-parent'); 

                element = document.querySelector(`[name=products]`);

                element?.setAttribute('class', ' de-activate nav-link Ripple-parent');
        }

        // set page's current button to activate
        // console.log("CLASSNAME:", className);
        element = document.querySelector(`[name=${btnElement}]`);
        element?.setAttribute('class', className + ' nav-link Ripple-parent');
};