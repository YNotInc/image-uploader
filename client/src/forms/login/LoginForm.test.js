import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import LoginForm from "./index";
import { api as API } from '../../utils/API';
import { act } from 'react-dom/test-utils'

describe("LoginForm", () => {
  // mock the response:
  const props = {
    getRole: jest.fn()
  };

  //mock api
  API.login = jest
    .fn()
    .mockImplementationOnce(() =>
    ({
      "data":
      {
        "message": "Auth successful", "access_token": "test", "refresh_token": "test", "expiration": "Sun Aug 22 2021 03:36:06 GMT+0000",
        "email": "admin@ymail.com"
      },
      "status": 200,
      "statusText": "OK",
      "request": {}
    }));


  it("updates the input value on change", () => {
    const { getByTestId } = render(<LoginForm {...props} />);
    // Grabs the test id
    const nameInput = getByTestId("login-email");
    fireEvent.change(nameInput, { target: { value: "test1@gmail.com" } });
    expect(nameInput.value).toBe("test1@gmail.com");
  });

  it("calls changeHandler on change", () => {
    const { getByLabelText } = render(<LoginForm {...props} />);
    // Grabs the input between the label for an input element
    const nameInput = getByLabelText("Email");
    fireEvent.change(nameInput, { target: { value: "test1@gmail.com" } });
    //Props no longer passed
    // expect(props.changeHandler).toHaveBeenCalled();
    expect(nameInput.value).toBe("test1@gmail.com");
  });

  it("should call submitHandler and axios when the form submits", async () => {

    // Arrange
    //mock api
    API.login = jest
      .fn()
      .mockImplementationOnce(() =>
      ({
        "data":
        {
          "message": "Auth successful", "access_token": "test", "refresh_token": "test", "expiration": "Sun Aug 22 2021 03:36:06 GMT+0000",
          "email": "admin@ymail.com"
        },
        "status": 200,
        "statusText": "OK",
        "request": {}
      }));

    // Target the form with Data test ID, to fire a submit event on the form
    let { getByTestId } = render(<LoginForm {...props} />);


    const form = getByTestId("login-form");
    // Act
    await act(async () => fireEvent.submit(form));

    // Assert: Anytime an action updates state it must be wrappred in act
    await act(async () => expect(API.login).toHaveBeenCalled());
    await act(async () => expect(props.getRole).toHaveBeenCalled());
  });
  it("should wipe the input fields when the form submits", async () => {

    // Arrange

    // Target the form with Data test ID, to fire a submit event on the form
    let { getByTestId, getByLabelText } = render(<LoginForm />);


    const form = getByTestId("login-form");
    // Act

    // ENter email and Grabs the input between the label for an input element
    const emailInput = getByLabelText("Email");
    // Since we are passing email from API.login, it is prefilled

    //wipe form data on submit:
    await act(async () => fireEvent.submit(form));

    await act(async () => expect(emailInput.value).toBe(""));
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<LoginForm {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});