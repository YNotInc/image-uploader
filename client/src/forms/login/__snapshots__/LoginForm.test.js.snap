// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="form-align flex-shrink-0"
    role="main"
  >
    <div
      class="container"
      data-test="container"
    >
      <div
        class="row"
        data-test="row"
      >
        <div
          class="col-12 align-self-center"
          data-test="col"
        >
          <form>
            <h1
              class="mt-5"
            >
              Login Form
            </h1>
            <br />
            <div
              class="form-group"
            >
              <div
                class="row"
                data-test="row"
              >
                <br />
              </div>
              <div
                class="row"
                data-test="row"
              >
                <div
                  class="col-auto"
                  data-test="col"
                >
                  <label
                    for="formGroupEmail"
                    id="formGroupEmail"
                  >
                    Email
                  </label>
                </div>
                <div
                  class="col"
                  data-test="col"
                >
                  <input
                    aria-labelledby="formGroupEmail"
                    class="form-control"
                    data-testid="login-email"
                    id="formGroupEmail"
                    name="email"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="row"
                data-test="row"
              >
                <br />
              </div>
              <div
                class="row"
                data-test="row"
              >
                <div
                  class="col-auto"
                  data-test="col"
                >
                  <label
                    for="formGroupPassword"
                    id="formGroupPassword"
                  >
                    Password
                  </label>
                </div>
                <div
                  class="col"
                  data-test="col"
                >
                  <input
                    aria-labelledby="formGroupPassword"
                    class="form-control"
                    data-testid="login-password"
                    id="formGroupPassword"
                    name="password"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <br />
              <div
                class="row"
                data-test="row"
              >
                <div
                  class="col"
                  data-test="col"
                >
                  <label
                    for="formGroupPassword"
                  />
                  <button
                    class="btn-blue-grey btn Ripple-parent mx-0"
                    data-test="button"
                    type="submit"
                  >
                    Submit
                    <div
                      class="Ripple "
                      data-test="waves"
                      style="top: 0px; left: 0px; width: 0px; height: 0px;"
                    />
                  </button>
                  <h3
                    class="mt-5"
                  />
                  <h3
                    class="mt-5"
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
